### PR TITLE
[TT-6419] Querying for __typename in any graph causes a 502 error

### PR DESCRIPTION
**Pull request by buraksezer**
This PR fixes an edge case in proxy-only mode. When you run the following query:

```graphql
query { __typename }
```

The GW panics and returns 502 for the request because the recent version of graphql-go-tools can handle this query without hitting the upstream. It maintains the type names internally. 

See https://github.com/TykTechnologies/graphql-go-tools/pull/323 for further info.

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4501/merge
Commit:  
Triggered by: pull_request (@buraksezer)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3639004576)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4501/merge
Commit: 9121c33da0d46c101b504e110baed3e5f2b78533 
Triggered by: pull_request (@buraksezer)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3639230879)

File: gateway/reverse_proxy.go [hunk 1]
```diff
@@ -1089,8 +1089,15 @@ func (p *ReverseProxy) handoverRequestToGraphQLExecutionEngine(roundTripper *Tyk
...
 		if isProxyOnly {
 			proxyOnlyCtx := reqCtx.(*GraphQLProxyOnlyContext)
-			header = proxyOnlyCtx.upstreamResponse.Header
-			httpStatus = proxyOnlyCtx.upstreamResponse.StatusCode
+			// There is a case in the proxy-only mode where the request can be handled
+			// by the library without calling the upstream.
+			// This is a valid query for proxy-only mode: query { __typename }
+			// In this case, upstreamResponse is nil.
+			// See TT-6419 for further info.
+			if proxyOnlyCtx.upstreamResponse != nil {
```
Comment (Aenimus) [inline gateway/reverse_proxy.go:1097]: Am I mistaken that there is a way to assign a variable but also check for equality in the same line?
Something like

if isProxyOnly && (proxyOnlyCtx := reqCtx.(*GraphQLProxyOnlyContext)) != nil {}

Basically an effort to reduce nesting? Maybe I am making things up. I tried to google, but I could have sworn I have seen it.

**Review by buger — APPROVED** [review]
Makes perfect sense 👌🏻

File: gateway/reverse_proxy.go (hunk 1 above)
Comment (buraksezer) [reply gateway/reverse_proxy.go:1097]: @Aenimus It's possible but I think this makes it harder to read. One level of nesting is not that bad.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4501/merge
Commit: f227ae41d2ea5b646557ab468126e79500228047 
Triggered by: pull_request (@buraksezer)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3655605454)

**Review by Aenimus — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: 6a32af5d5b5457426ffd961f3be6c12d9d671f6e [TT-6419] Querying for __typename in any graph causes a 502 error (#4501)

This PR fixes an edge case in proxy-only mode. When you run the
following query:

```graphql
query { __typename }
```

The GW panics and returns 502 for the request because the recent version
of graphql-go-tools can handle this query without hitting the upstream.
It maintains the type names internally.

See https://github.com/TykTechnologies/graphql-go-tools/pull/323 for
further info.
Triggered by: push (@buraksezer)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3656410516)

**Pull request merged**